crypto/bn/bn_prime.c

```diff
@@ -458,16 +420,23 @@ static int probable_prime(BIGNUM *rnd, int bits, prime_t *mods)
     return 1;
 }
 
-int bn_probable_prime_dh(BIGNUM *rnd, int bits,
-                         const BIGNUM *add, const BIGNUM *rem, BN_CTX *ctx)
+static int probable_prime_dh(BIGNUM *rnd, int bits, int safe, prime_t *mods,
+                             const BIGNUM *add, const BIGNUM *rem,
+                             BN_CTX *ctx)
 {
     int i, ret = 0;
     BIGNUM *t1;
+    BN_ULONG delta;
+    BN_ULONG maxdelta = BN_MASK2 - primes[NUMPRIMES - 1];
 
     BN_CTX_start(ctx);
     if ((t1 = BN_CTX_get(ctx)) == NULL)
         goto err;
 
+    if (maxdelta > BN_MASK2 - BN_get_word(add))
+        maxdelta = BN_MASK2 - BN_get_word(add);
```

Comment:
It should also subtract `primes[NUMPRIMES - 1]`

Reply:
maxdelta is the minimum of `maxdelta = BN_MASK2 - primes[NUMPRIMES - 1]`
 and `BN_MASK2 - BN_get_word(add))`

Reply:
therefore neither `mods[i] + delta` nor `delta += BN_get_word(add)` will overflow.

Reply:
did I overlook something?

Reply:
I think the whole `delta`/`maxdelta` needs a review. In `mods` you have numbers in the range `0 - primes[NUMPRIMES-1]`. We need to make sure that adding delta to that doesn't overflow. We also need to make sure that adding `add` to `delta` doesn't overflow.

Reply:
(github didn't show any of your comments before I replied.)

After looking more careful, it seems we skip the delta that we could theoretically do, but would cause an overflow trying to get to the next one.

Reply:
actually the range of mods is `0 .. primes[NUMPRIMES-1] - 1`